src/services/goToDefinition.ts

```diff
@@ -180,7 +180,7 @@ export function getDefinitionAtPosition(program: Program, sourceFile: SourceFile
     if (!symbol && isModuleSpecifierLike(fallbackNode)) {
         // We couldn't resolve the module specifier as an external module, but it could
         // be that module resolution succeeded but the target was not a module.
-        const ref = program.getResolvedModule(sourceFile, fallbackNode.text, getModeForUsageLocation(sourceFile, fallbackNode))?.resolvedModule;
+        const ref = program.getResolvedModule(sourceFile, fallbackNode.text, getModeForUsageLocation(sourceFile, fallbackNode, program.getCompilerOptions()))?.resolvedModule;

```

Comment:
And here

I feel like we can add internal API to program that gives the mode from sourceFile and literal and handles referenced project options

Reply:
I think it needs to be public (and maybe even deprecate the existing `getModeForUsageLocation`), otherwise this will be a footgun for API users.

Reply:
In most of the cases the options supplied to this should have been the redirect options .. (eg.  CompilerHost.getSourceFile should pass correct options to the "sourceFile" creation but clearly thats incorrect today.